imap/imapparse.c

```diff
@@ -259,160 +259,112 @@ EXPORTED int getxstring(struct protstream *pin, struct protstream *pout,
     return EOF;
 }
 
-EXPORTED int getint32(struct protstream *pin, int32_t *num)
+/* can't flag with -1 if there is no number here, so return EOF */
+static int _getint(struct protstream *pin, uint64_t max, uint64_t *num)
 {
-    int32_t result = 0;
+    uint64_t result = 0;
+    uint64_t tens = max / 10;
+    uint8_t ones = max % 10;
     int c = EOF;
     int gotchar = 0;
 
-    /* INT_MAX == 2147483647 */
     while ((c = prot_getc(pin)) != EOF && cyrus_isdigit(c)) {
-        if (result > 214748364 || (result == 214748364 && (c > '7')))
+        uint8_t digit = c - '0';
+
+        if (result > tens || (result == tens && (digit > ones)))
             fatal("num too big", EX_IOERR);
-        result = result * 10 + c - '0';
+        result = result * 10 + digit;
         gotchar = 1;
     }
 
+    *num = result;
+
     if (!gotchar) {
         if (c != EOF) prot_ungetc(c, pin);
         return EOF;
     }
 
-    *num = result;
-
     return c;
 }
 
-/* Like getint32() but explicitly signed, i.e. negative numbers
- * are accepted */
-EXPORTED int getsint32(struct protstream *pin, int32_t *num)
+/* Like _getint() but explicitly signed, i.e. negative numbers are accepted */
+static int _getsint(struct protstream *pin, uint64_t max, int64_t *num)
 {
     int c;
     int sgn = 1;
+    uint64_t result;
 
     c = prot_getc(pin);
-    if (c == EOF)
+    if (c == EOF) {
+        *num = 0;
         return EOF;
+    }
 
-    if (c == '-')
+    if (c == '-') {
         sgn = -1;
-    else if (c == '+')
-        sgn = 1;
-    else
+        max++; // min = (-max - 1)
+    }
+    else if (c != '+') {
         prot_ungetc(c, pin);
+    }
+
+    c = _getint(pin, max, &result);
+
+    *num = sgn * result;
 
-    c = getint32(pin, num);
-    if (c == EOF)
-        return EOF;
-    /* this is slightly buggy: the number INT_MIN = -2147483648
-     * is a valid signed 32b int but is not accepted */
-    if (sgn < 0)
-        *num = - (*num);
     return c;
 }
 
-/* can't flag with -1 if there is no number here, so return EOF */
-EXPORTED int getuint32(struct protstream *pin, uint32_t *num)
+EXPORTED int getint32(struct protstream *pin, int32_t *num)
 {
-    uint32_t result = 0;
-    int c = EOF;
-    int gotchar = 0;
-
-    /* UINT_MAX == 4294967295U */
-    while ((c = prot_getc(pin)) != EOF && cyrus_isdigit(c)) {
-        if (result > 429496729 || (result == 429496729 && (c > '5')))
-            fatal("num too big", EX_IOERR);
-        result = result * 10 + c - '0';
-        gotchar = 1;
-    }
-
-    if (!gotchar) {
-        if (c != EOF) prot_ungetc(c, pin);
-        return EOF;
-    }
+    uint64_t result;
+    int c = _getint(pin, INT_MAX, &result);
 
     *num = result;
```

Comment:
The really big thing for me here is these `*num = result` assignments, when num is of a narrower type with a different signedness.  Does it work?  I dunno, probably, mostly.  But this sort of conversion is where C can be tricky to reason about, and even if it's technically good by the book, does the compiler get it right? Does the optimiser? Etc.

This is where, for the sake of avoiding having to think about (and second-guess) the integer conversion, I'd kinda rather just keep the separate implementations.